src/entity.ts

```diff
@@ -130,86 +128,119 @@ export class Entity {
 		Property$init(prop, this, value);
 	}
 
+	withContext(context: InitializationContext, action: (entity: Entity) => void) {
+		// Don't overwrite existing context
+		if (!this._context)
+			this._context = context;
+		// Ensure provided context waits on the existing context to be ready
+		else
+			context.wait(new Promise(resolve => this._context.ready(resolve)));
+
+		action(this);
+
+		if (context !== null) {
+			context.ready(() => {
+				this._context = null;
+			});
+		}
+	}
+
 	set(properties: ObjectLookup<any>): void;
 	set(property: string, value: any): void;
 	set(property: any, value?: any): void {
 		let properties: ObjectLookup<any>;
 
 		// Convert property/value pair to a property dictionary
-		if (typeof property === "string") {
-			properties = {};
-			properties[property] = value;
-		}
-		else {
+		if (typeof property === "string")
+			properties = { [property]: value };
+		else
 			properties = property;
-		}
 
 		// Set the specified properties
 		for (let [propName, state] of Entity.getSortedPropertyData(properties)) {
 			const prop = this.serializer.resolveProperty(this, propName);
 			if (prop) {
-				let value;
-				const currentValue = prop.value(this);
-				if (isEntityType(prop.propertyType)) {
-					const ChildEntity = prop.propertyType;
-					if (prop.isList && Array.isArray(state) && Array.isArray(currentValue)) {
-						state.forEach((s, idx) => {
-							if (!(s instanceof ChildEntity))
-								s = this.serializer.deserialize(this, s, prop, null, false);
-
-							// Modifying/replacing existing list item
-							if (idx < currentValue.length) {
-								// If the item is an object that has an Id property, then retrieve or create an object with that Id
-								if (!(s instanceof ChildEntity) && typeof s === "object" && getIdFromState(ChildEntity.meta, s))
-									s = ChildEntity.meta.createSync(s);
-								if (s instanceof ChildEntity)
-									state.splice(idx, 1, s);
-								else
-									currentValue[idx].set(s);
-							}
-							// Add a list item
-							else if (s instanceof ChildEntity)
-								currentValue.push(s);
-							else
-								currentValue.push(ChildEntity.meta.createSync(s));
-						});
+				if (this._context) {
```

Comment:
What do you think about getting rid of the nested `if` and `continue` and making it an `if`/`else`, using a ternary (like `valueResolution = this._context ? this._context.tryResolveValue(this, prop, state) : null`)? Could be just my personal preference, but I think code that uses `break`/`continue` can be a bit harder to understand, and may be more susceptible to refactor bugs. I don't have any data to back that up, just a hunch.

Reply:
Yeah, I didn't have a great feeling about the use of `continue` here. I'll rework that.

Reply:
Fixed.